File: src/server/crdt/generic/dwflag.go
package crdt

import (
	"fmt"
	g01 "sdle-server/proto"
)

type DWFlag struct {
	replicaID string
	dotKernel *DotKernel[struct{}]
}

func NewDWFlag(replicaID string) *DWFlag {
	return &DWFlag{
		replicaID: replicaID,
		dotKernel: NewDotKernel[struct{}](),
	}
}

func (flag *DWFlag) Context() *DotContext {
	return flag.dotKernel.Context()
}

func (flag *DWFlag) SetContext(ctx *DotContext) {
	flag.dotKernel.SetContext(ctx)
}

func (flag *DWFlag) Read() bool {
	return len(flag.dotKernel.dotValues) == 0
}

func (flag *DWFlag) Disable() *DWFlag {
	delta := NewDWFlag(flag.replicaID)

	delta.dotKernel = flag.dotKernel.Reset()
	delta.dotKernel.Join(flag.dotKernel.Add(flag.replicaID, struct{}{}))

	return delta
}

func (flag *DWFlag) Enable() *DWFlag {
	delta := NewDWFlag(flag.replicaID)

	delta.dotKernel = flag.dotKernel.Reset()

	return delta
}

func (flag *DWFlag) Reset() *DWFlag {
	delta := NewDWFlag(flag.replicaID)
	delta.dotKernel = flag.dotKernel.Reset()
	return delta
}

func (flag *DWFlag) Join(other *DWFlag) {
	flag.dotKernel.Join(other.dotKernel)
}

func (flag *DWFlag) IsNull() bool {
	return len(flag.dotKernel.dotValues) == 0
}

func (flag *DWFlag) Clone() *DWFlag {
	clone := NewDWFlag(flag.replicaID)
	clone.dotKernel = flag.dotKernel.Clone()
	return clone
}

func (flag *DWFlag) String() string {
	return fmt.Sprintf("EWFlag{replicaID: %s, dotKernel: %v}", flag.replicaID, flag.dotKernel)
}

func (flag *DWFlag) ToProto() *g01.DWFlag {
	return &g01.DWFlag{
		DotKernel: (*EmptyDotKernel)(flag.dotKernel).ToProto(),
	}
}

func DWFlagFromProto(protoFlag *g01.DWFlag, replicaId string, ctx *DotContext) *DWFlag {
	dotKernel := EmptyDotKernelFromProto(protoFlag.GetDotKernel(), ctx)

	return &DWFlag{
		replicaID: replicaId,
		dotKernel: (*DotKernel[struct{}])(dotKernel),
	}
}

File: src/client/types/crdt/generic/dwflag.ts
import { isIterableEmpty } from "@/lib/utils";
import DotContext from "./dot-context";
import DotKernel from "./dot-kernel";
import { EmptyDotKernel, DWFlag as DWFlagProto } from "@/lib/proto/client";

export default class DWFlag {
    replicaId: string;
    dotKernel: DotKernel<null>;

    constructor(replicaId: string) {
        this.replicaId = replicaId;
        this.dotKernel = new DotKernel<null>();
    }

    public getContext() {
        return this.dotKernel.getContext();
    }

    public setContext(context: DotContext): void {
        this.dotKernel.setContext(context);
    }

    public read(): boolean {
        return isIterableEmpty(this.dotKernel.getValues());
    }

    public disable(): DWFlag {
        const delta = new DWFlag(this.replicaId);

        delta.dotKernel = this.dotKernel.reset();
        delta.dotKernel.join(this.dotKernel.add(this.replicaId, null));

        return delta;
    }

    public enable(): DWFlag {
        const delta = new DWFlag(this.replicaId);

        delta.dotKernel = this.dotKernel.reset();

        return delta;
    }

    public join(other: DWFlag): void {
        this.dotKernel.join(other.dotKernel);
    }

    public isNull(): boolean {
        return isIterableEmpty(this.dotKernel.getValues());
    }

    public clone(): DWFlag {
        const clone = new DWFlag(this.replicaId);
        clone.dotKernel = this.dotKernel.clone();
        return clone;
    }

    public toProto(): DWFlagProto {
        const stringDotKernel = this.dotKernel.toEmptyProto();
        return DWFlagProto.create({
            dotKernel: stringDotKernel,
        });
    }

    public static fromProto(proto: DWFlagProto, replicaId: string): DWFlag {
        const dotKernel = DotKernel.fromEmptyProto(proto.dotKernel as EmptyDotKernel, new DotContext());

        const flag = new DWFlag(replicaId);
        flag.dotKernel = dotKernel;

        return flag;
    }
}